src/win.c

```diff
@@ -320,11 +345,81 @@ void win_release_images(struct backend_base *backend, struct managed_win *w) {
 	}
 }
 
+/// Returns true if the `prop` property is stale, as well as clears the stale flag.
+static bool win_fetch_and_unset_property_stale(struct managed_win *w, xcb_atom_t prop);
+/// Returns true if any of the properties are stale, as well as clear all the stale flags.
+static bool win_check_and_clear_all_properties_stale(struct managed_win *w);
+
 void win_process_update_flags(session_t *ps, struct managed_win *w) {
 	if (win_check_flags_all(w, WIN_FLAGS_MAPPED)) {
 		map_win_start(ps, w);
 		win_clear_flags(w, WIN_FLAGS_MAPPED);
 	}
+
+	// Check client first, because later property updates need accurate client window
+	// information
+	if (win_check_flags_all(w, WIN_FLAGS_CLIENT_STALE)) {
+		win_recheck_client(ps, w);
+		win_clear_flags(w, WIN_FLAGS_CLIENT_STALE);
+	}
```

Comment:
Doing this check/update here should fix a "rare" race-condition when a client window disappears and the frame is still rendered in the next cycle, but it's shadow has been invalidated due to specific shadow-rules (see #479). Should fix the crashes described in #390 and #487.

Reply:
OK, can I merge this first, and you rebase #479 ? Or do you want that to be merged first?

Reply:
Merge this first. I'm gona rebase #479 with the mentioned comments. Doing the check/update here seem more plausible.